### PR TITLE
(Not)Be(Abstract|Sealed|Static) class assertions implemented

### PR DESCRIPTION
**Pull request by melchiork**
For  #645
Assertions for types (abstract, sealed, static) added. They are based on C# understanding of those keywords, as in CLR they differ.

### CODE REVIEW

**Comment by melchiork** [comment]
@jnyrup Thanks, fixed. Also I've added two examples to the docs.

**Comment by jnyrup** [comment]
It just struck me that from a C# perspective the keywords `sealed`, `abstract` and `static` are only applicable to class like types, and not interfaces or structs.

Several assertions in `TypeAssertions` checks if the type is an interface.
One example is `BeDerivedFrom` as you cannot derive from an interface type, but implement it.

**Comment by dennisdoomen** [comment]
So? You want to have some changea made to this PR?

**Comment by jnyrup** [comment]
Yes, I would like to have a check that the `Subject` type is not a valuetype or an interface or throw an `InvalidOperationException` with a descriptive message.

This is similar to what we do in `ActionAssertions.FailIfSubjectIsAsyncVoid`, `XDocumentAssertions.HaveRoot` and `XDocumentAssertions.HaveElement`

Calling `typeof(StructType).IsSealed` returns `true`, but that is from a CLR perspective.
The purpose of the type assertions seems to be from a C# perspective, i.e. if the type under assertion has the `sealed` keyword.

**Comment by melchiork** [comment]
@jnyrup @dennisdoomen 
Done, I've also check if the type is a Delegate.
I went with ArgumentException as similar was done in TypeAssertions.Implement and TypeAssertions.NotImplement

**Comment by jnyrup** [comment]
`BeDerivedFrom` and the alike uses `ArgumentException` as they're checking the argument. 
As the `Subject` isn't an argument to the assertion method I prefer `InvalidOperationException` like in `HaveRoot`. 

I really like your use of `Theory` tests. 
Excellent job!

**Comment by melchiork** [comment]
@jnyrup Done. Thanks for the reviews!

File: Src/FluentAssertions/Types/TypeAssertions.cs [hunk 1]
```diff
@@ -1102,5 +1216,14 @@ protected override string Identifier
         {
             get { return "type"; }
         }
+
+        private void AssertThatSubjectIsClass()
+        {
+            var typeInfo = Subject.GetTypeInfo();
+            if (typeInfo.IsInterface || typeInfo.IsValueType || typeof(Delegate).IsAssignableFrom(typeInfo.BaseType))
+            {
+               throw new InvalidOperationException($"{Subject} must be a class.");
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/TypeAssertions.cs]: Last thing I can come up with is a missing space to keep four space indentation.

**Pull request merged**